contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,302 @@
+        return _revokeRoles(resource, role, account);
+    }
+
+
+    /**
+     * @dev Revokes `role` from the calling account.
+     *
+     * Roles are often managed via {grantRole} and {revokeRole}: this function's
+     * purpose is to provide a mechanism for accounts to lose their privileges
+     * if they are compromised (such as when a trusted device is misplaced).
+     *
+     * If the calling account had been revoked `role`, emits a {RoleRevoked}
+     * event.
+     *
+     * Requirements:
+     *
+     * - the caller must be `callerConfirmation`.
+     *
+     * May emit a {RoleRevoked} event.
+        *
+     * Returns `true` if the role was revoked, `false` otherwise.
+     */
+    function renounceRole(bytes32 resource, uint256 role, address callerConfirmation) public virtual returns (bool) {
+        if (callerConfirmation != _msgSender()) {
+            revert EnhancedAccessControlBadConfirmation();
+        }
+
+        return _revokeRoles(resource, role, callerConfirmation);
+    }
+
+    // Internal functions
+
+    /**
+     * @dev Reverts with an {AccessControlUnauthorizedAccount} error if `account`
+     * is missing `role`.
+     */
+    function _checkRole(bytes32 resource, uint256 role, address account) internal view virtual {
+        if (!hasRoles(resource, role, account)) {
+            revert EnhancedAccessControlUnauthorizedAccountRole(resource, role, account);
+        }
+    }
+
+    /**
+     * @dev Reverts with an {AccessControlUnauthorizedAccount} error if account does not have the admin role for the given role.
+     */
+    function _checkCanGrantRole(bytes32 resource, uint256 role, address account) internal view virtual {
+        uint256 theAdminRole = getRoleAdmin(role);
+        if (!hasRoles(resource, theAdminRole, account)) {
+            if (!hasRoles(resource, DEFAULT_ADMIN_ROLE, account)) {
+                revert EnhancedAccessControlUnauthorizedAccountAdminRole(resource, role, account);
+            }
+        }
+    }
+
+    /**
+     * @dev Locks roles within a resource to prevent their removal.
+     * Adds to any existing locked roles.
+     *
+     * @param resource The resource to lock roles in.
+     * @param roleBitmap The bitmap of roles to lock.
+     */
+    function _lockRoles(bytes32 resource, uint256 roleBitmap) internal virtual {
+        _lockedRoles[resource] |= roleBitmap;
+        emit EnhancedAccessControlRolesLocked(resource, roleBitmap);
+    }
+
+    /**
+     * @dev Copies all roles from `srcAccount` in `srcResource` to `dstAccount` in `dstResource`.
+     *
+     * @param srcResource The resource to copy roles from.
+     * @param srcAccount The account to copy roles from.
+     * @param dstResource The resource to copy roles to.
+     * @param dstAccount The account to copy roles to.
+     */
+    function _copyRoles(bytes32 srcResource, address srcAccount, bytes32 dstResource, address dstAccount) internal virtual {
```

Comment:
Does it make sense to copy roles across resources? I think this should probably be limited to copying them within a single resource.